prereise/gather/hiflddata/plant_agg.py

```diff
@@ -0,0 +1,153 @@
+        Pmin[str(row["Plant Name"]).upper()] = row["Minimum Load (MW)"]
+    return Pmin
+
+#Get the lattitude and longitude of plants 
+def Loc_of_plant():
+    loc_of_plant = {}
+    csv_data = pd.read_csv("data/Power_Plants.csv")
+    for index, row in csv_data.iterrows():
+        loc = (format(row['LATITUDE'], coord_precision),format(row['LONGITUDE'], coord_precision))
+        loc_of_plant[row['NAME']] = loc
+    return loc_of_plant
+
+def Plant_agg(clean_data,ZipOfsub_dict,loc_of_plant,LocOfsub_dict,Pmin):
+    plant_dict = {}
+  
+    for index, row in clean_data.iterrows():
+        tu = (row['PLANT'],row['TYPE'])
+        if(tu not in plant_dict):
+            if(row['PLANT'] in Pmin and row['PLANT'] in loc_of_plant):
+                if(row['ZIP'] in ZipOfsub_dict):
+                    min_d = 100000.0
+                    min_s = ""
+                    for value in ZipOfsub_dict[row['ZIP']]:
+                    # calculate the distance between the plant and substations
+                        if(geodesic(loc_of_plant[row['PLANT']],LocOfsub_dict[value]).m < min_d):
+                            min_s = value
+                            min_d = geodesic(loc_of_plant[row['PLANT']],LocOfsub_dict[value]).m
+                    bus_id = value
+                    pmin = Pmin[row['PLANT']]
+            # if this zip does not contain subs, we try to find subs in neighbor zip.
+                else: 
+                    zi = int(row['ZIP'])
+                    for i in range(-5,6):
+                        min_d = 100000.0
+                        min_s = ""
+                        if(str(zi+i) in Pmin and str(zi+i) in loc_of_plant):
+                            for value in ZipOfsub_dict[str(zi+i)]:
+                                if(geodesic(loc_of_plant[row['PLANT']],LocOfsub_dict[value]).m < min_d):
+                                    min_s = value
+                                    min_d = geodesic(loc_of_plant[row['PLANT']],LocOfsub_dict[value]).m
+                    bus_id = value
+                    pmin = 0
+            pmaxwin = row['WINTER_CAP']
+            pmaxsum = row['SUMMER_CAP']
+            list1 = [bus_id,pmaxwin,pmaxsum,pmin]
+            plant_dict[tu] = list1
+        else:
+            list1 = plant_dict[tu]
+            list1[1] = list1[1] + row['WINTER_CAP']
+            list1[2] = list1[2] + row['SUMMER_CAP']
+            plant_dict[tu] = list1
+    return plant_dict
+
+def write_plant(plant_dict):
+    plant = open('output/plant.csv','w',newline='')
+    csv_writer = csv.writer(plant)
+    csv_writer.writerow(["plant_id","bus_id","Pg","status","Pmax","Pmin","ramp_30","type"])
+    for key in plant_dict:
+        list1 = plant_dict[key]
+        csv_writer.writerow([key[0]+'-'+key[1],list1[0], 1, "OP", min(list1[1],list1[2]),list1[3],list1[3],key[1]])
+    plant.close()
+
+def write_gen(plant_dict, type_dict):
+    gencost = open('output/gencost.csv','w',newline='')
```

Comment:
We should use the `with open(...) as gencost:` syntax here to make sure that we always close the file, even if there is some bug before we get to the `gencost.close()` line.

Reply:
Thanks. All the file open blocks have been updated to use `with open`